velox/expression/Expr.cpp

```diff
@@ -417,17 +417,29 @@ bool Expr::checkGetSharedSubexprValues(
   if (!rows.isSubset(*sharedSubexprRows_)) {
     LocalSelectivityVector missingRowsHolder(context, rows);
     auto missingRows = missingRowsHolder.get();
-    assert(missingRows); // lint
+    VELOX_DCHECK(missingRows); // lint
     missingRows->deselect(*sharedSubexprRows_);
 
-    // Fix finalSelection at "rows" if missingRows is a strict subset to avoid
-    // losing values outside of missingRows.
-    bool updateFinalSelection = context.isFinalSelection() &&
-        (missingRows->countSelected() < rows.countSelected());
-    VarSetter finalSelectionOr(
-        context.mutableFinalSelection(), &rows, updateFinalSelection);
-    VarSetter isFinalSelectionOr(
-        context.mutableIsFinalSelection(), false, updateFinalSelection);
+    // Add the missingRows to sharedSubexprRows_ that will eventually be
+    // evaluated and added to sharedSubexprValues_.
+    sharedSubexprRows_->select(*missingRows);
+
+    // Fix finalSelection to avoid losing values outside missingRows.
+    LocalSelectivityVector newFinalSelectionHolder(
+        context, *sharedSubexprRows_);
+    auto newFinalSelection = newFinalSelectionHolder.get();
+    VELOX_DCHECK(newFinalSelection); // lint
+    if (!context.isFinalSelection()) {
+      // In case currently set finalSelection does not include all rows in
+      // sharedSubexprRows_.
+      VELOX_DCHECK(context.finalSelection() != nullptr);
+      newFinalSelection->select(*context.finalSelection());
+    }
+    VarSetter finalSelectionPreservePrecomputedValues(
+        context.mutableFinalSelection(),
+        const_cast<const SelectivityVector*>(newFinalSelection));
+    VarSetter isFinalSelectionPreservePrecomputedValues(
+        context.mutableIsFinalSelection(), false, context.isFinalSelection());
```

Comment:
The previous line updates context.finalSelection unconditionally. This line updates context.isFinalSelection only if context.isFinalSelection() is true. Shouldn't both of these updates happen together unconditionally?

Reply:
For the second line we are attempting to set isFinalSelection() to false, so if it is already set to false then we dont need to update it, regardless of the condition, isFinalSelection() after that line would be false so keeping the condition or setting it unconditionally will be the same.
since both are same, either will work, let me know if you prefer removing the condition.

Also after #2399 is merged I can update this to use the new class ScopedFinalSelectionSetter added there.

Reply:
Got it. Thank you for explaining. Either way is fine with me. Using ScopedFinalSelectionSetter will make it easier to read I think.